core/parachain/validator/impl/parachain_processor.cpp

```diff
@@ -7,38 +7,615 @@
+      logger_->debug("Statement of {} already issued.", candidate_hash);
+      return;
+    }
+
+    appendAsyncValidationTask(id, relay_parent, peer_id);
+  }
+
+  void ParachainProcessorImpl::requestPoV() {
+    logger_->error("Not implemented.");
+  }
+
+  void ParachainProcessorImpl::kickOffValidationWork(
+      AttestingData &attesting_data) {
+    /// TODO(iceseer): add to awaiting_validation to be sure no doublicated
+    /// validations
+
+    auto opt_descriptor = candidateDescriptorFrom(attesting_data.statement);
+    BOOST_ASSERT(opt_descriptor);
+
+    auto const &collator_id = collatorIdFromDescriptor(*opt_descriptor);
+    if (our_current_state_.required_collator
+        && collator_id != *our_current_state_.required_collator) {
+      our_current_state_.issued_statements.insert(
+          attesting_data.candidate_hash);
+      return;
+    }
+
+    notify_internal(context_, [wptr{weak_from_this()}] {
+      if (auto self = wptr.lock()) {
+        /// request PoV
+        self->requestPoV();
+        /// validate PoV
+      }
+    });
+  }
+
+  void ParachainProcessorImpl::appendAsyncValidationTask(
+      network::ParachainId id,
+      primitives::BlockHash const &relay_parent,
+      libp2p::peer::PeerId const &peer_id) {
+    /// TODO(iceseer): add to awaiting_validation to be sure no doublicated
+    /// validations
+    notify_internal(
+        context_, [id, relay_parent, peer_id, wptr{weak_from_this()}] {
+          if (auto self = wptr.lock()) {
+            self->logger_->debug(
+                "Got an async task for VALIDATION execution {}:{}:{}",
+                id,
+                relay_parent,
+                peer_id.toBase58());
+
+            if (auto collation_result =
+                    self->getFromSlot(id, relay_parent, peer_id))
+              self->validateAndMakeAvailable(
+                  peer_id,
+                  relay_parent,
+                  std::move(collation_result).value(),
+                  [wptr](auto const &peer_id,
+                         auto &&validate_and_second_result) {
+                    if (auto self = wptr.lock())
+                      self->onValidationComplete(
+                          peer_id, std::move(validate_and_second_result));
+                  });
+            else
+              self->logger_->template warn(
+                  "In job retrieve collation result error: {}",
+                  collation_result.error().message());
+          }
+        });
+  }
+
+  template <typename T>
+  outcome::result<network::Signature> ParachainProcessorImpl::sign(
+      T const &t) const {
+    if (!keypair_) return Error::KEY_NOT_PRESENT;
+
+    auto payload = scale::encode(t).value();
+    return crypto_provider_->sign(*keypair_, payload).value();
+  }
+
+  std::optional<network::ValidatorIndex> ParachainProcessorImpl::getOurIndex() {
+    logger_->error("Not implemented. Return my validator index.");
+    return std::nullopt;
+  }
```

Comment:
May use `ValidatorSigner` from https://github.com/soramitsu/kagome/pull/1336 (TODO: signable trait or conversion function, e.g. bitfield is encoded as is, but statement has prefix and conversion to other type)